CommonTools/RecoAlgos/plugins/SimClusterRecHitAssocitionProducer.cc

```diff
@@ -0,0 +1,136 @@
+#include "SimDataFormats/CaloAnalysis/interface/SimCluster.h"
+#include "DataFormats/CaloRecHit/interface/CaloRecHit.h"
+#include "SimDataFormats/CaloHit/interface/PCaloHit.h"
+#include "SimDataFormats/CaloHit/interface/PCaloHitContainer.h"
+
+#include "DataFormats/Common/interface/Association.h"
+
+#include "FWCore/Utilities/interface/transform.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include <set>
+
+//
+// class decleration
+//
+typedef std::pair<size_t, float> IdxAndFraction;
+
+class SimClusterRecHitAssociationProducer : public edm::stream::EDProducer<> {
+public:
+  explicit SimClusterRecHitAssociationProducer(const edm::ParameterSet&);
+  ~SimClusterRecHitAssociationProducer() override;
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  std::vector<edm::InputTag> caloRechitTags_;
+  std::vector<edm::EDGetTokenT<edm::PCaloHitContainer>> caloSimhitCollectionTokens_;
+  //std::vector<edm::EDGetTokenT<std::vector<PSimHit>> trackSimhitCollectionTokens_;
+  std::vector<edm::EDGetTokenT<edm::View<CaloRecHit>>> caloRechitCollectionTokens_;
+  edm::EDGetTokenT<SimClusterCollection> scCollectionToken_;
+};
+
+SimClusterRecHitAssociationProducer::SimClusterRecHitAssociationProducer(const edm::ParameterSet& pset)
+    :  //caloSimhitCollectionTokens_(edm::vector_transform(pset.getParameter<std::vector<edm::InputTag>>("caloSimHits"),
+      //  [this](const edm::InputTag& tag) {return mayConsume<edm::PCaloHitContainer>(tag); })),
+      //trackSimhitCollectionTokens_(edm::vector_transform(pset.getParameter<edm::InputTag>("trackSimHits"),
+      //    [this](const edm::InputTag& tag) {return mayConsume<std::vector<PSimHit>(tag); }),
+      caloRechitTags_(pset.getParameter<std::vector<edm::InputTag>>("caloRecHits")),
+      caloRechitCollectionTokens_(edm::vector_transform(
+          caloRechitTags_, [this](const edm::InputTag& tag) { return mayConsume<edm::View<CaloRecHit>>(tag); })),
+      scCollectionToken_(consumes<SimClusterCollection>(pset.getParameter<edm::InputTag>("simClusters"))) {
+  for (auto& tag : caloRechitTags_) {
+    const std::string& label = tag.instance();
+    produces<edm::Association<SimClusterCollection>>(label + "ToSimClus");
+  }
+  produces<std::unordered_map<int, float>>();
+}
+
+SimClusterRecHitAssociationProducer::~SimClusterRecHitAssociationProducer() {}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void SimClusterRecHitAssociationProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  auto simClusterToRecEnergy = std::make_unique<std::unordered_map<int, float>>();
+  edm::Handle<SimClusterCollection> scCollection;
+  iEvent.getByToken(scCollectionToken_, scCollection);
+  std::unordered_map<size_t, IdxAndFraction> hitDetIdToIndex;
+  std::unordered_map<size_t, float> hitDetIdToTotalSimFrac;
+
+  for (size_t s = 0; s < scCollection->size(); s++) {
+    const auto& sc = scCollection->at(s);
```

Comment:
after the change proposed above
```suggestion
    const auto& sc = scCollection[s];
```
similarly, in other places do not use `at(` if the range is known to be correct